Add bus factor analyses

christinaexyou opened the pull request:
Quantifies the amount of contributors a project can afford to lose before it stalls by hypothetically having these people get run over a bus. Here contributors include, but are not limited to: 1) individuals who've made commits, 2) created issues, and 3) creating, reviewing, and engaging in conversation around PRs.

We further parameterized the general notion of bus factor by allowing the user to set a custom threshold for the percent of contributors a project can afford to lose, to count outliers, and to analyze how it evolves over the lifetime of the project in sliding time windows and step sizes.

All analyses in this notebook are done using Ansible as an example but are replicable with any other GitHub repository.

JamesKunstle commented on notebooks/bus-factor.ipynb:
```diff
@@ -0,0 +1,1606 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #8.        filter = df['email'].str.contains('bot')
</pre><p>Email could have 'bot' in it, or the username might include 'bot'. Ideally we'd also parameterize this function so that we could scrub out known bot names (in some repos, bots operate under standard user accounts of have a funny name like "Borg" or something).</p><p> </p><p>i.e. def remove_bots(df, regex_pattern, columns=["email"]):</p><p> </p><p>so that we can screen based on some pattern and pick the columns that we screen out.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

JamesKunstle replied:
<p>Would love to have a pattern to screen out AND a list of names that are also screened out- if someone knows all of the bots in their repo/org they can provide it.</p>

JamesKunstle commented on notebooks/bus-factor.ipynb:
```diff
@@ -0,0 +1,1606 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #11.        df = df[(np.abs(stats.zscore(df[activity])) &lt; 2)]
</pre><p>you're clobbering the non-outlier df here, not sure if you're meaning to do that.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

cdolfi replied:
<p>A little more context here would be great im a little lost on this step</p>

JamesKunstle replied:
<p>Z-score captures # of standard deviation a value is from the mean. If the absolute value of the Z-score is less than 2, it's not an outlier. The line in question only accepts outliers, discarding non-outliers.</p>

christinaexyou replied:
<p>I changed this line so we only filter out for upper bound outliers and I added comment explaining what it does.</p>

JamesKunstle replied:
Okay cool

JamesKunstle commented on notebooks/bus-factor.ipynb:
```diff
@@ -0,0 +1,1606 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #27.        for i in range(len(df)):
</pre><p>Should note in the docs that this function assumes that you've sorted the input in descending order. I know you're doing that in 'preprocess_df' but please also parameterize the function to sort the input. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

christinaexyou replied:
<p>I added a parameter ascending and by default it is set to False. </p>

JamesKunstle commented on notebooks/bus-factor.ipynb:
```diff
@@ -0,0 +1,1606 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #1.    bf_by_windows = bus_factor_by_windows(conn_commits, activity='commit',start_date='2017-01-24', end_date= '2018-01-24', window_width=6, step_size=4, threshold=0.5)
</pre><p>Let's show from the beginning of the project to now. I'm more interested in that, because it can show that a project is more sustainable as it gets bigger.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

JamesKunstle replied:
<p>I'd like to see 'bus_factor_by_windows' applied here too.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='28'/>

JamesKunstle commented:
@christinaexyou Three over-arching comments:
(1) This is becoming an excellent notebook, I'm very excited for this to be integrated into 8K. 
(2) I made some parameterization requests concerning the 'bot' in email thing. I'd like to be able to test for username/email patterns anywhere, if possible.
(3) I'd love to see 'bus_factor_by_windows' over the lifetime of the project graphed for each of the perspectives.

If you finish those things we'll review once more for correctness and have Cali + Oindrilla + Hema take a look as well, and beyond that this is a great start to this problem!

christinaexyou commented:
In addition to making the above edits, I added a section titled `Conclusion` where I compare the bus factor across all the different perspectives I analyzed.

cdolfi commented on notebooks/bus-factor.ipynb:
```diff
@@ -0,0 +1,2312 @@
+{
```
<p>I see the value in identifying the outliers that make contributions significantly above, but what is the value of knowing which contributors are significantly below? Also for this notebook it might be good to have some cell(s) that identify which contributors are the outliers/ bus factor but to not run them when you push. Curious to hear @Jameskunstle thoughts on this</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

JamesKunstle replied:
<p>Yeah personally I'm only interested in those above. There really isn't an outlier lower-bound, someone might make a single issue ever and that's a reasonable amount of contribution.</p>

cdolfi replied:
<p>Remove "below average" from description </p>

cdolfi commented on notebooks/bus-factor.ipynb:
```diff
@@ -0,0 +1,2312 @@
+{
```
<p>This use of a pie chart might be a bit misleading. Pie charts are visualizing the parts that make up a complete space, while this is showing the percent of the activity (commits, prs, etc) done by each of the top 10 with the total space being the activity of the top 10 people. Whats the goal of this visualization/what information do you want people to walk away from? Could see this being displayed in a different style graph or showing it where the total pie chart is the all the activity (commits, prs, etc whichever graph it is) and then the actions done not by the top 10 is just grouped into an "other" category. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='60'/>

JamesKunstle replied:
<p>Perfect feedback. I'd  like to have all of the other contribs as a section of the pie chart that I can screen out if I like (to show the scale of the highest impact contribs). could also do a stacked bar graph per category.</p>

christinaexyou replied:
<img width="661" alt="Screenshot 2023-07-23 at 7 06 46 PM" src="https://github.com/oss-aspen/Rappel/assets/77992688/d28cd14b-a09f-4ced-a8e0-b6db9170239e">

cdolfi commented:
@christinaexyou This is looking awesome! Awesome work. I put a few comments in. The big thing I left with trying to think about the best way to visualize this information. A good way to think about it is thinking about what information or perspective you want the user to gain and then consider which type of visualization will give that. I really like the idea of atleast one implementation of it being a pie chart where the total space is all commits (as an example), the "bus factor" individuals are pieces of the pie and then everything done not by them is grouped together as "other".

JamesKunstle commented on notebooks/bus-factor.ipynb:
```diff
@@ -0,0 +1,2312 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #2.    with open('.././env.list', 'r') as file:
</pre><p>We should make it easier for contributors. Ideally, I'd recommend using environment variables, but we also gave the parameters to you as a newline-separated list of values. We should clean that up in the future. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

JamesKunstle replied:
<pre class="ql-syntax" spellcheck="false">Line #5.        returns a DataFrame where the columns are contrb_id, their commit, and the commit date 
</pre><p>doesn't always return a commit, does it? Just executes a query against augur and returns a df.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

JamesKunstle replied:
<pre class="ql-syntax" spellcheck="false">Line #7.        conn = pd.read_sql_query(repo_query, con=engine.connect())
</pre><p>variable should be 'df' rather than conn</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

JamesKunstle replied:
<pre class="ql-syntax" spellcheck="false">Line #1.    def remove_bots(df, keyword='bot', colname='email'):
</pre><p>I think this function can be refactored to:</p><pre class="ql-syntax" spellcheck="false">def screen_column(df, patterns=["bot"], col="email"):
  :param df: DataFrame to screen column values out of.
  :param patterns: list of strings or regular expression, patterns to be screened out
  :param col: name of col to screen patterns out of
  
  returns a pd.DataFrame with columns that matched, and
  column used to match, dropped.
  """
    Creates a filter of col values that match at least one
    entry in patterns. Negates filter, drops matching rows.
    Drops screening column.
    https://pandas.pydata.org/docs/reference/api/pandas.Series.str.match.html
  """      
  return df[
    ~df[col].str.match(patterns, case=False)
  ].drop(colname, axis=1)
      
</pre><p> </p><p>This lets us match for regular expressions, or for regular strings. might need to use str.fullmatch(), unsure.</p><p>Also no sure about the indentation / having the dataframe brackets off-sync like that.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

JamesKunstle replied:
<pre class="ql-syntax" spellcheck="false">Line #1.    def preprocess_df(df, activity, keyword='bot', colname='email', ascending=False):
</pre><p> </p><p>rename 'activity' to 'contribution_type'</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

JamesKunstle replied:
<pre class="ql-syntax" spellcheck="false">Line #17.        # sort rows according to amount of activity in descending order
</pre><p>order, descending/ascending is parameterized so it's not always descending.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

JamesKunstle replied:
<pre class="ql-syntax" spellcheck="false">Line #3.        :param df: A DataFrame in which the number of contributions are sorted in descending order
</pre><p>"A DataFrame of contributor_id and contribution_number rows, sorted in descending order of contribution_number."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

JamesKunstle replied:
<pre class="ql-syntax" spellcheck="false">Line #6.        :param remove_outliers: A boolean indicating whether to remove outliers or not
</pre><p>not using this parameter.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

JamesKunstle replied:
<pre class="ql-syntax" spellcheck="false">Line #14.        # calculate total number of commits
</pre><p>not always considering commits- I believe that you reuse this function for issues and PRs too.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

JamesKunstle replied:
<pre class="ql-syntax" spellcheck="false">Line #23.        df = df.assign(cntrbs = lambda x: (x[activity]/t_activity))
</pre><p>I think this could be rewritten as:</p><p> </p><p>df["cntrbs"] = df[activity] / t_activity</p><p> </p><p>because the division will be vectorized. This way we don't have to use a map function implicitly, and can use the vectorized interface.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

JamesKunstle replied:
<pre class="ql-syntax" spellcheck="false">Line #27.        for i in range(len(df)):
</pre><p> </p><p>Here's one way that you can make this logic a little easier to read:</p><p>docs: https://sparkbyexamples.com/pandas/iterate-over-rows-in-pandas-dataframe/#:~:text=Using%20DataFrame.,()%20to%20Iterate%20Over%20Rows&amp;text=iterrows()%20is%20used%20to,row%5B%22Fee%22%5D%20.</p><p> </p><pre class="ql-syntax" spellcheck="false">for _, row in df.iterrows():
	cum_per += row["cntrbs"]
	contrib_ids.append(row["cntrb_id"])
	if cum_per &gt;= threshold:
		break
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

JamesKunstle replied:
<pre class="ql-syntax" spellcheck="false">Line #5.        :param step_size:  An integer representing the sliding window in months
</pre><p>"how many months the sliding window advances each iteration"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

JamesKunstle replied:
<p>Since we're tracking the bus factor over time, I think I'd prefer to use a line-graph.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

JamesKunstle replied:
<p>I don't think we can "expect" at least two outliers. The top contributor has 2x the number of commits as the third, which is pretty reasonable. It's really hard to intuit outliers (1) from intuition and (2) without seeing all of the data.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

JamesKunstle replied:
<p>This is another graph I'm having trouble believing. Ansible is a big project, to have only one or two people account for 50%+ of PR's is really strange to me.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='42'/>

JamesKunstle commented on notebooks/bus-factor.ipynb:
```diff
@@ -0,0 +1,2312 @@
+{
```
<p>I'm having trouble believing this graph. I don't expect the bus factor of ansible to be only 1 from 2020 -&gt; 2022. Maybe a bot is doing the committing? idk. Should look at the raw data and try to sus that out.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

cdolfi replied:
<p>^^ agreed</p>

JamesKunstle commented on notebooks/bus-factor.ipynb:
```diff
@@ -0,0 +1,2312 @@
+{
```
<p>this doesn't make sense, there should never be a busfactor of 0, the minimum would be exactly 1 when only 1 person does n% of the work. I think this is because there are not outlier PR reviewers, and we're screening out all non-outliers in the beginning.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='48'/>

christinaexyou replied:
<p>I checked the returned df from the SQL query for PR reviews for 2020 and it seems like there aren't any so the root cause is some error in my query statement.</p>

christinaexyou replied:
<p>Fixed the error in my SQL query - selected the wrong timestamp.</p>

christinaexyou replied:
<img width="1063" alt="Screenshot 2023-07-23 at 7 05 13 PM" src="https://github.com/oss-aspen/Rappel/assets/77992688/232b6cf4-3d6c-4d66-bf93-027e34ac8a03">

sgoggins replied:
When I think of bus factor, I think PR reviews give you a distinct measure of how many maintainers there are. There can be significant committers with a larger n, and often there is, FWIW.

christinaexyou commented:
Edits made:
1. Minor changes in code as suggested by James
2. Corrected an error in my SQL query for PRs reviewed
3. Changed my visualization framework to Plotly 
4. Replaced bar plots by line graphs
5. Changed my pie charts to include ALL contributions - the top 10 are explicitly labelled with their cntrb_ids and the remaining as other

Next steps:
- Validate my SQL queries, particularly for PRs

JamesKunstle commented on notebooks/bus-factor.ipynb:
```diff
@@ -0,0 +1,8845 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #24.        # iterate through each row in df and create a running sum of percent contributions
</pre><p>Add this comment:</p><p>"Contributor is ranked by how large of a proportion of all work they did- if someone authored 350 of 1000 commits, for instance, they'll account for 35% of the commits made. The list is ranked in descending order. Ordered list is greedily processed such that the smallest list of contributors is found that accounts for <em>threshold</em> percent of the work done for a given contribution type." </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

cdolfi replied:
<p>@jameskunstle should this be a comment or a markdown cell? </p>

JamesKunstle commented on notebooks/bus-factor.ipynb:
```diff
@@ -0,0 +1,8845 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #38.            cntrbs = preprocess_df(df.loc[mask], contribution_type)
</pre><p>also, I'd rename this variable. it doesn't mean a lot semantically.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

christinaexyou replied:
<p>Renamed to count_cntrbs_window.</p>

JamesKunstle commented on notebooks/bus-factor.ipynb:
```diff
@@ -0,0 +1,8845 @@
+{
```
<p>Here's my plain english interpretation of this query:</p><p> </p><p>"List every distinct commit hash (uniquely identifying individual commits) in the repo along with the Augur ID of the contributor who authored the commit, the email of the committer of the commit, and the date that the commit was authored. If a committer uses more than one email, make sure that the ID linked to the commit is the highest-level contributor ID possible- the owner of all of the emails that are aliases."</p><p> </p><p>One problem here is that you're using 'cmt_committer_email' but considering 'cmt_author_date' which might be for two different people, because the author and the committer can be different. You should choose either the author or the committer (I prefer the author because they're more likely to have made a code contribution, imo).</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

christinaexyou replied:
<p>Replaced the committer email with the author email.</p>

JamesKunstle commented on notebooks/bus-factor.ipynb:
```diff
@@ -0,0 +1,8845 @@
+{
```
<p>Plain english interpretation:</p><p>"For all issues in the repo, get the distinct ID of each issue, the ID of the contributor who reported the issue, the email of that contributor, and the date that the issue was created."</p><p> </p><p>What I don't know is whether ca.canonical_email is doing the right thing based on what we're trying to do. Since a user has multiple emails, I assume that they're canonical email is... something specific? Maybe the one most directly linked to the used account? In any case, there might be a more appropriate field for that.</p><p> </p><p>I'm also not sure about i.reporter_id. It could be exactly right, I just need to read the schema.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='29'/>

christinaexyou replied:
<p>Agreed, I'm not sure what the difference is between canonical vs alias email since they're usually the same. </p><p> </p>

cdolfi replied:
I am pretty sure canonical is the first email hit from a github account and alias is any email associated with the account. ill get @sgoggins to confirm

sgoggins replied:
I am going to look at the code to determine canonical_email/reporter_id, cntrb_id questions. 

The tl;dr on various data elements is: 
issues: reporter_id
commits: depends on context and criticality of recency. I will look at how aliases are, or are not addressed. If we are using email, then the aliases are kind of important because everybody gets one row in the contributor table, and nearly everyone has 3 or more rows in the alias table. I think I have used 15 different emails for commits over the past 300 years, for example. 
prs: cntrb_id

Looking at code now.

JamesKunstle commented on notebooks/bus-factor.ipynb:
```diff
@@ -0,0 +1,8845 @@
+{
```
<p>Here's my plain english interpretation:</p><p>"For a given repo, get a list of each PR including that PR's ID, the Augur ID of the contributor who opened the PR, the email of the contributor who authored the PR, and the date that the PR was opened, where the PR opener was not also the reviewer and the PR has been merged."</p><p> </p><p>This is another query that I have to spend a bit of time on. I'm not sure how to interpret pull_request_events or pull_request_reviewers well. We can talk on Thursday.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='37'/>

cdolfi replied:
<p>If this pr is just for looking at who and when it is created, why do you need pull_request_events and reviewers? I feel like this can be based on just the pr and repo table. This might solve the issue below with the 2017 problem. There is definitely pr data for 2017 in ansible/ ansible. Another thing to note is the problem here might also be that there is not always a reviewer, especially in ansible/ansible </p><p> </p>

christinaexyou replied:
<p>Good point, I wasn't sure if the pr_augur_contributor_id column in pull_requests represented PR creators or reviewers but my line graph makes alot more sense now.</p>

JamesKunstle commented on notebooks/bus-factor.ipynb:
```diff
@@ -0,0 +1,8845 @@
+{
```
<p>I don't think this query is correct- the 'pr' column is non-unique, and merging a PR should be a unique event. We'll look at this.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='43'/>

christinaexyou replied:
<p>Since we're looking at contributors that had their PRs merged, my logic was that if multiple people collaborated on a commit we would get non-unique values for the 'pr' and 'date' column.</p>

cdolfi replied:
@sgoggins check this one out as well

JamesKunstle commented on notebooks/bus-factor.ipynb:
```diff
@@ -0,0 +1,8845 @@
+{
```
<p>It's also weird to think that between 1 and 4 people are consistently handling &gt;50% of the comments on the project, but I guess that makes sense.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='53'/>

JamesKunstle commented:
@christinaexyou I think this is getting super close to being done. I made a few small requests but once we do our PR overview and @cdolfi looks it over again I think we'll be able to merge and work on moving this into an 8Knot page. Superb work, thanks for all of your time.

cdolfi commented:
@christinaexyou  Completely agree with @JamesKunstle this looks great! Most of the comments I would have made James would have beat me too it. Please feel free to reach out to me if you would like any help or guidance at the 8Knot integration step :)

christinaexyou commented:
@JamesKunstle @cdolfi 
Edits made:

1.  Renamed functions for more clarity
2. Replaced canonical email with alias email
3. SQL statements to fetch PR info
** there are duplicate pull_request_id's and cntrb_id's in the returned PR reviewers df since PR reviewers can contributed to a single PR in multiple ways as defined by the pull_request_events table i.e merging and closing PRs

JamesKunstle commented on notebooks/bus-factor.ipynb:
```diff
@@ -0,0 +1,8937 @@
+{
```
<p>" with commits that are 2 standard deviations or more above the mean number of commits." 'commits' to 'contributions' or something, just being generic in the comment.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

christinaexyou replied:
<p>Replaced 'commits' with 'contributions'.</p>

JamesKunstle commented on notebooks/bus-factor.ipynb:
```diff
@@ -0,0 +1,8937 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #32.        while start_date + window_width &lt; end_date:
</pre><p>Realized that this might not get the last window, which would be in range, if 'start + window = end_date + 0.000001' or something. Would make sense as a do-while look if that was in Python. Maybe a while-loop with this condition as the break-out or something?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

JamesKunstle replied:
<p>No longer considering only those PRs that have been created and also merged- only considering created.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='40'/>

JamesKunstle commented on notebooks/bus-factor.ipynb:
```diff
@@ -0,0 +1,8937 @@
+{
```
<p>Not sure how the bus factor is 0 for 6 months in 2017. There were hundred of PRs merged in 2017 in ansible/ansible. We should look at the raw data and try to understand this. Could be that that table is incomplete in Augur as well.</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/oss-aspen/Rappel/pull/220/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='41'/>

christinaexyou replied:
<p>When I filter for rows in the data frame where the year is 2017, I get an empty set.</p>

JamesKunstle commented:
@christinaexyou 
Tiny adjustments, and one data question, then I'm done reviewing lol.

christinaexyou commented:
@JamesKunstle @cdolfi 

Edits made:
1. Fixed my SQL queries for PR info 
2. `bus_factor_by_windows` calculates the bus factor for the remaining slice of time in the data frame, even if it is less than the window width

JamesKunstle reviewed (approved):
I'm done reviewing.
I feel very confident in this analysis as it exists, and any further correctness-checking will be done in audits during integration into 8Knot.
A lingering question I have is about PR review query correctness- will submit a PR myself fixing if it's identified as a problem.
@christinaexyou thank you so much for working on this, I know the review process has been long but, from my perspective, the confidence I now have in this work is significant.

cdolfi commented:
@JamesKunstle If you are on board I can merge this. @christinaexyou just want to send additional kudos of your work here. The in-depth review process can be uncomfortable and daunting, especially when you are new to the process. This high level of work is of quality of an associate, not just of an intern which is something I hope you are proud of because I sure am.

JamesKunstle commented:
++ agreed @cdolfi
